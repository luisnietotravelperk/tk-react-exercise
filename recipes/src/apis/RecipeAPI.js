const baseUrl = 'http://localhost:8000/api/recipe/recipe/'

const _requestFetch = (url, options = {}, returnData = true) => {
  const abortController = new AbortController()
  const fetchOptions = {
      method: 'GET',
      headers: { "Content-Type": "application/json" },
      signal: abortController.signal
  }

  Object.keys(options).map(key => fetchOptions[key] = options[key])

  return fetch(url, fetchOptions)
    .then(res => {
      if (!res.ok) {
        throw Error('Could not complete the operation')
      }

      if (returnData) {
        return res.json()
      } else {
        return true
      }
    })
    .catch(e => {
      if (e.name !== 'AbortError') {
        throw e
      }
    })
}

export const getRecipes = (name)  => {
  let url = baseUrl
  if (name) {
    url += `?name=${name}`
  }

  return _requestFetch(url)
}

export const getRecipe = (id) => {
  let url = `${baseUrl}${id}/`
  return _requestFetch(url)
}

export const createRecipe = (data) => {
  let url = baseUrl
  let options = {
    method: 'POST',
    body: JSON.stringify(data)
  }

  return _requestFetch(url, options)
}

export const updateRecipe = (id, data) => {
  let url = `${baseUrl}${id}/`
  let options = {
    method: 'PUT',
    body: JSON.stringify(data)
  }

  return _requestFetch(url, options)
}

export const deleteRecipe = (id) => {
  let url = `${baseUrl}${id}/`
  return _requestFetch(url, { method: 'DELETE'}, false)
}
