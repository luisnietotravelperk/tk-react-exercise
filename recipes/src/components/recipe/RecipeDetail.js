import { Link, useParams } from "react-router-dom";
import useFetch from "../../hooks/useFetch";
import { Button, ButtonBox, MainSection } from "../Common";

const RecipeDetail = () => {
  let params = useParams()

  // Get the information of the recipe
  const baseUrl = `http://localhost:8000/api/recipe/recipe/${params.recipeId}/`
  const { data: recipe, isPending, error } = useFetch(baseUrl)

  return (
    <MainSection>
      { error && <div>{error}</div>}
      { isPending && <div>Loading...</div>}
      { recipe && (
        <section>
          <h1>{recipe.name}</h1>
          <h2>Description</h2>
          <p>{recipe.description}</p>
          <h2>Ingredients</h2>
          { recipe.ingredients.length > 0 && recipe.ingredients.map(m => (
            <li key={m.id}>{m.name}</li>
          )) }
          { recipe.ingredients.length === 0 && (
            <li>Not ingredients were register</li>
          )}
          <ButtonBox>
            <Button>Delete</Button>
            <Link to="/recipes">
              <Button>Back</Button>
            </Link>
          </ButtonBox>
        </section>
      )}
    </MainSection>
  );
}

export default RecipeDetail;