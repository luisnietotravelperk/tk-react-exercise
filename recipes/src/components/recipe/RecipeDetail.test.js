import { render, screen } from "@testing-library/react";
import RecipeDetail from "./RecipeDetail";
import { getRecipe } from '../../apis/RecipeAPI'
import { BrowserRouter, useParams } from "react-router-dom";


jest.mock('../../apis/RecipeAPI')
jest.mock("react-router-dom", () => ({
 ...jest.requireActual("react-router-dom"),
 useParams: jest.fn(),
}));

describe('<RecipeDetail />', () => {
  it('Render all the information', async () => {
    const recipe = {
      id: 1,
      name: 'recipe 1',
      description: 'Lorem ipsum',
      ingredients: [
        { id: 1, name: 'ingredient 1' },
        { id: 2, name: 'ingredient 2' },
        { id: 3, name: 'ingredient 3' },
      ]
    }
    useParams.mockReturnValue({ recipeId: recipe.id })
    getRecipe.mockResolvedValueOnce(Promise.resolve(recipe))

    const { asFragment } = render(
      <BrowserRouter>
        <RecipeDetail />
      </BrowserRouter>
    )

    await screen.findByText('recipe 1')

    expect(getRecipe).toHaveBeenCalledTimes(1)
    expect(getRecipe).toHaveBeenCalledWith(recipe.id)
    expect(asFragment()).toMatchSnapshot()
  })
})