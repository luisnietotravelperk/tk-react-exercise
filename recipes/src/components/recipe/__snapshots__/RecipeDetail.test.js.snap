// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeDetail /> Render all the component 1`] = `
<DocumentFragment>
  <article
    class="sc-dkzDqf glcQRH"
  >
    <section>
      <h1>
        recipe 1
      </h1>
      <h2>
        Description [
        <a
          href="/recipes/update/1"
        >
          Edit
        </a>
        ]
      </h2>
      <p>
        Lorem ipsum
      </p>
      <h2>
        Ingredients
      </h2>
      <li>
        ingredient 1
      </li>
      <li>
        ingredient 2
      </li>
      <li>
        ingredient 3
      </li>
      <div
        class="sc-bczRLJ bNBYBW"
      >
        <button
          class="sc-gsnTZi ciHGB"
        >
          Delete
        </button>
        <a
          href="/recipes"
        >
          <button
            class="sc-gsnTZi ciHGB"
          >
            Back
          </button>
        </a>
      </div>
    </section>
  </article>
</DocumentFragment>
`;

exports[`<RecipeDetail /> Render all the information 1`] = `
<DocumentFragment>
  <article
    class="sc-dkzDqf glcQRH"
  >
    <section>
      <h1>
        recipe 1
      </h1>
      <h2>
        Description [
        <a
          href="/recipes/update/1"
        >
          Edit
        </a>
        ]
      </h2>
      <p>
        Lorem ipsum
      </p>
      <h2>
        Ingredients
      </h2>
      <li>
        ingredient 1
      </li>
      <li>
        ingredient 2
      </li>
      <li>
        ingredient 3
      </li>
      <div
        class="sc-bczRLJ bNBYBW"
      >
        <button
          class="sc-gsnTZi ciHGB"
        >
          Delete
        </button>
        <a
          href="/recipes"
        >
          <button
            class="sc-gsnTZi ciHGB"
          >
            Back
          </button>
        </a>
      </div>
    </section>
  </article>
</DocumentFragment>
`;
